File: Sources/HabibiFloat/BubbleSoundPlayer.swift
import AppKit
import Foundation

enum BubbleSoundPreset: String, CaseIterable {
    case waterDrop
    case softBloop
    case jellyPop
    case budak
    case bubbleChime
    case muted

    var title: String {
        switch self {
        case .waterDrop: "Water Drop"
        case .softBloop: "Soft Bloop"
        case .jellyPop: "Jelly Pop"
        case .budak: "Budak"
        case .bubbleChime: "Bubble Chime"
        case .muted: "No Sound"
        }
    }

    static func from(id: String) -> BubbleSoundPreset {
        BubbleSoundPreset(rawValue: id) ?? .waterDrop
    }
}

enum BubbleSoundVolume: String, CaseIterable {
    case soft
    case normal
    case loud

    var title: String {
        switch self {
        case .soft: "Soft"
        case .normal: "Normal"
        case .loud: "Loud"
        }
    }

    var multiplier: Float {
        switch self {
        case .soft: 0.55
        case .normal: 1.0
        case .loud: 1.55
        }
    }

    static func from(id: String) -> BubbleSoundVolume {
        BubbleSoundVolume(rawValue: id) ?? .normal
    }
}

final class BubbleSoundPlayer {
    private var soundBanks: [BubbleSoundPreset: [NSSound]] = [:]
    private var indexes: [BubbleSoundPreset: Int] = [:]

    init() {
        for preset in BubbleSoundPreset.allCases where preset != .muted {
            let data = Self.makeBubbleWav(for: preset)
            let sounds = (0..<4).compactMap { _ in NSSound(data: data) }
            sounds.forEach { sound in
                sound.volume = volume(for: preset)
            }
            soundBanks[preset] = sounds
        }
    }

    func play(presetID: String, volumeID: String) {
        let preset = BubbleSoundPreset.from(id: presetID)
        guard preset != .muted, let sounds = soundBanks[preset], !sounds.isEmpty else {
            return
        }

        let index = indexes[preset, default: 0]
        let sound = sounds[index]
        indexes[preset] = (index + 1) % sounds.count
        sound.volume = min(volume(for: preset) * BubbleSoundVolume.from(id: volumeID).multiplier, 1)
        sound.stop()
        sound.currentTime = 0
        sound.play()
    }

    private func volume(for preset: BubbleSoundPreset) -> Float {
        switch preset {
        case .waterDrop: 0.34
        case .softBloop: 0.28
        case .jellyPop: 0.36
        case .budak: 0.38
        case .bubbleChime: 0.30
        case .muted: 0
        }
    }

    private static func makeBubbleWav(for preset: BubbleSoundPreset) -> Data {
        let sampleRate = 44_100
        let duration = 0.42
        let sampleCount = Int(Double(sampleRate) * duration)
        var samples = [Int16]()
        samples.reserveCapacity(sampleCount)

        for frame in 0..<sampleCount {
            let t = Double(frame) / Double(sampleRate)
            let value = signal(for: preset, at: t)
            let softened = tanh(value) * 0.64
            samples.append(Int16(max(-1, min(1, softened)) * Double(Int16.max)))
        }

        var data = Data()
        data.appendASCII("RIFF")
        data.appendUInt32LE(UInt32(36 + samples.count * 2))
        data.appendASCII("WAVE")
        data.appendASCII("fmt ")
        data.appendUInt32LE(16)
        data.appendUInt16LE(1)
        data.appendUInt16LE(1)
        data.appendUInt32LE(UInt32(sampleRate))
        data.appendUInt32LE(UInt32(sampleRate * 2))
        data.appendUInt16LE(2)
        data.appendUInt16LE(16)
        data.appendASCII("data")
        data.appendUInt32LE(UInt32(samples.count * 2))

        for sample in samples {
            data.appendUInt16LE(UInt16(bitPattern: sample))
        }

        return data
    }

    private static func signal(for preset: BubbleSoundPreset, at t: Double) -> Double {
        switch preset {
        case .waterDrop:
            droplet(t, start: 0.00, duration: 0.13, startFrequency: 610, endFrequency: 980, gain: 0.74) +
                droplet(t, start: 0.08, duration: 0.16, startFrequency: 420, endFrequency: 760, gain: 0.42) +
                droplet(t, start: 0.19, duration: 0.12, startFrequency: 760, endFrequency: 1180, gain: 0.28)
        case .softBloop:
            droplet(t, start: 0.00, duration: 0.20, startFrequency: 280, endFrequency: 430, gain: 0.70) +
                droplet(t, start: 0.14, duration: 0.15, startFrequency: 330, endFrequency: 520, gain: 0.32)
        case .jellyPop:
            droplet(t, start: 0.00, duration: 0.09, startFrequency: 340, endFrequency: 220, gain: 0.90) +
                droplet(t, start: 0.05, duration: 0.13, startFrequency: 720, endFrequency: 980, gain: 0.40)
        case .budak:
            droplet(t, start: 0.00, duration: 0.07, startFrequency: 190, endFrequency: 140, gain: 1.00) +
                droplet(t, start: 0.045, duration: 0.10, startFrequency: 520, endFrequency: 360, gain: 0.54) +
                droplet(t, start: 0.12, duration: 0.10, startFrequency: 760, endFrequency: 900, gain: 0.26)
        case .bubbleChime:
            droplet(t, start: 0.00, duration: 0.17, startFrequency: 820, endFrequency: 1220, gain: 0.56) +
                droplet(t, start: 0.11, duration: 0.20, startFrequency: 1040, endFrequency: 1460, gain: 0.34)
        case .muted:
            0
        }
    }

    private static func droplet(
        _ t: Double,
        start: Double,
        duration: Double,
        startFrequency: Double,
        endFrequency: Double,
        gain: Double
    ) -> Double {
        guard t >= start, t <= start + duration else {
            return 0
        }

        let progress = (t - start) / duration
        let frequency = startFrequency + (endFrequency - startFrequency) * progress
        let envelope = sin(progress * .pi) * exp(-progress * 2.35)
        let wobble = sin(2 * .pi * 9 * progress) * 0.025
        return sin(2 * .pi * frequency * (t - start) + wobble) * envelope * gain
    }
}

private extension Data {
    mutating func appendASCII(_ string: String) {
        append(contentsOf: string.utf8)
    }

    mutating func appendUInt16LE(_ value: UInt16) {
        append(UInt8(value & 0xff))
        append(UInt8((value >> 8) & 0xff))
    }

    mutating func appendUInt32LE(_ value: UInt32) {
        append(UInt8(value & 0xff))
        append(UInt8((value >> 8) & 0xff))
        append(UInt8((value >> 16) & 0xff))
        append(UInt8((value >> 24) & 0xff))
    }
}

File: windows/Bubbly.Windows/Services/BubbleSoundPlayer.cs
using System;
using System.IO;
using System.Media;
using System.Threading.Tasks;

namespace Bubbly.Windows.Services;

public sealed class BubbleSoundPlayer
{
    public void Play(string presetId, string volumeId)
    {
        if (presetId == "muted")
        {
            return;
        }

        try
        {
            var data = MakeBubbleWav(presetId, VolumeMultiplier(volumeId));
            _ = Task.Run(() =>
            {
                using var stream = new MemoryStream(data);
                using var player = new SoundPlayer(stream);
                player.PlaySync();
            });
        }
        catch
        {
            // Sound should never block the companion interaction.
        }
    }

    private static double VolumeMultiplier(string volumeId)
    {
        return volumeId switch
        {
            "soft" => 0.55,
            "loud" => 1.55,
            _ => 1.0
        };
    }

    private static byte[] MakeBubbleWav(string presetId, double volumeMultiplier)
    {
        const int sampleRate = 44100;
        const double duration = 0.42;
        var sampleCount = (int)(sampleRate * duration);
        using var stream = new MemoryStream();
        using var writer = new BinaryWriter(stream);

        writer.Write("RIFF"u8.ToArray());
        writer.Write(36 + sampleCount * 2);
        writer.Write("WAVE"u8.ToArray());
        writer.Write("fmt "u8.ToArray());
        writer.Write(16);
        writer.Write((short)1);
        writer.Write((short)1);
        writer.Write(sampleRate);
        writer.Write(sampleRate * 2);
        writer.Write((short)2);
        writer.Write((short)16);
        writer.Write("data"u8.ToArray());
        writer.Write(sampleCount * 2);

        for (var frame = 0; frame < sampleCount; frame++)
        {
            var t = (double)frame / sampleRate;
            var softened = Math.Tanh(Signal(presetId, t)) * 0.64 * volumeMultiplier;
            writer.Write((short)(Math.Clamp(softened, -1, 1) * short.MaxValue));
        }

        return stream.ToArray();
    }

    private static double Signal(string presetId, double t)
    {
        return presetId switch
        {
            "softBloop" =>
                Droplet(t, 0.00, 0.20, 280, 430, 0.70) +
                Droplet(t, 0.14, 0.15, 330, 520, 0.32),
            "jellyPop" =>
                Droplet(t, 0.00, 0.09, 340, 220, 0.90) +
                Droplet(t, 0.05, 0.13, 720, 980, 0.40),
            "budak" =>
                Droplet(t, 0.00, 0.07, 190, 140, 1.00) +
                Droplet(t, 0.045, 0.10, 520, 360, 0.54) +
                Droplet(t, 0.12, 0.10, 760, 900, 0.26),
            "bubbleChime" =>
                Droplet(t, 0.00, 0.17, 820, 1220, 0.56) +
                Droplet(t, 0.11, 0.20, 1040, 1460, 0.34),
            _ =>
                Droplet(t, 0.00, 0.13, 610, 980, 0.74) +
                Droplet(t, 0.08, 0.16, 420, 760, 0.42) +
                Droplet(t, 0.19, 0.12, 760, 1180, 0.28)
        };
    }

    private static double Droplet(double t, double start, double duration, double startFrequency, double endFrequency, double gain)
    {
        if (t < start || t > start + duration)
        {
            return 0;
        }

        var progress = (t - start) / duration;
        var frequency = startFrequency + (endFrequency - startFrequency) * progress;
        var envelope = Math.Sin(progress * Math.PI) * Math.Exp(-progress * 2.35);
        var wobble = Math.Sin(2 * Math.PI * 9 * progress) * 0.025;
        return Math.Sin(2 * Math.PI * frequency * (t - start) + wobble) * envelope * gain;
    }
}
